### DISCUSSION TITLE
Dynamic field selection

### DISCUSSION THREAD

**Discussion by arthurgailes** [prisma/prisma#15988] [Q&A] [answered]
Hello,

Is there a way to use select with a variable while maintaining type safety? I have a function like this:
```
const variable = input.variable as keyof Prisma.MyTableSelect;

const data = prisma.MyTable.findMany({
  where: { id },
  select: { id: true, [variable]: true },
  // include: {}, TODO: add places
});
```

But when I use the dynamic variable, it destroys my type definition, converting it to `PrismaPromise<{}[]>`. What I'd like to have it revert to is something like:

```
PrismaPromise<{
    id: string;
    var1?: number;
    var2?: number;
}[]>
```

Any help would be greatly appreciated.

**Comment by nurul3101** [comment, marked as answer]
Hello @arthurgailes 👋 

I don't think you could achieve this natively yet, and we do have a Proposal for [Client Extension](https://github.com/prisma/prisma/issues/15074) which would allow you to achieve this in future. In the Proposal, we have a section for extending your queries which would be helpful in your case. We are working on getting this shipped in future Prisma releases.
For now, you could use [Raw Query](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access) to inject dynamic field names. Do let me know if you face any issues in achieving this through Raw Query.

**Reply by anitacung** [reply]
👋Hi @arthurgailes,
 
I'm going through pending GitHub Discussions to check if you need any help. I see that Nurul has given you a suggestion to use [Raw Query](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access). 

Is this issue still open?

**Reply by arthurgailes** [reply]
Thank you @nurul3101. I don't think the Raw Query method works. From the page: "Template variables can only be used for data values (such as email in the example above). **Variables cannot be used for identifiers such as column names**, table names or database names, or for SQL keywords."

**Reply by nurul3101** [reply]
Hey @arthurgailes, thanks for getting back to me, what I meant was to use Raw Query Methods and not particularly `$queryRaw`. As you mentioned if you would use `$queryRaw` then injecting dynamic table/columns names would not be possible, so specifically, you would need to use the `$queryRawUnsafe` method to achieve injecting dynamic table/column names. Here's an [example](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#dynamic-table-names-in-postgresql) of injecting dynamic table name through `$queryRawUnsafe`. 
Please do let me know in case you face any issues with this approach. 🙂

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
